packages/main/src/plugin/kubernetes-client.ts

```diff
@@ -1049,64 +1010,159 @@ export class KubernetesClient {
+  }
+
+  /**
+   * Apply a given yaml file to a context, i.e. 'kubectl apply -f'. Resources that exist
+   * on the context are patched, and any new resources are created.
+   *
+   * @param context a context
+   * @param filePath file system path to a YAML Kubernetes spec
+   * @param namespace the namespace to use for any resources that don't include one
+   * @return an array of resources created
+   */
+  async applyResourcesFromFile(context: string, filePath: string, namespace?: string): Promise<KubernetesObject[]> {
+    const manifests = await this.loadManifestsFromFile(filePath);
+    return this.syncResources(context, manifests, 'apply', namespace);
+  }
+
+  /**
+   * Apply a given yaml file to a context, i.e. 'kubectl apply -f'. Resources that exist
+   * on the context are patched, and any new resources are created.
+   *
+   * @param context a context
+   * @param filePath file system path to a YAML Kubernetes spec
+   * @param namespace the namespace to use for any resources that don't include one
+   * @return an array of resources created
+   */
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  async applyResources(context: string, manifests: any[], namespace?: string): Promise<KubernetesObject[]> {
+    return this.syncResources(context, manifests, 'apply', namespace);
+  }
+
+  /**
+   * Applies or creates a set of resources to a context, via creation or patching.
+   *
+   * @param context a context
+   * @param manifests a set of Kubernetes spec manifests
+   * @param action 'create' (only create new resources, do not overwrite existing) or 'apply' (create or patch as necessary)
+   * @param namespace the namespace to use for any resources that don't include one
+   * @return an array of resources created
+   *
+   * Heavily influenced by the API example:
+   * https://github.com/kubernetes-client/javascript/blob/0fbfd8fc2dcc7f4ec3e6fcd64a5c55169b6ef0b8/examples/typescript/apply/apply-example.ts
+   */
+
+  async syncResources(
+    context: string,
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    manifests: any[],
+    action: 'create' | 'apply',
+    namespace?: string,
+  ): Promise<KubernetesObject[]> {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    const telemetryOptions: Record<string, any> = {
+      manifestsSize: manifests?.length,
+      action: action,
+    };
+    if (namespace) {
+      telemetryOptions.namespace = namespace;
+    }
+
     try {
       const ctx = new KubeConfig();
       ctx.loadFromFile(this.kubeconfigPath);
       ctx.currentContext = context;
-      for (const manifest of manifests) {
-        // https://github.com/kubernetes-client/javascript/issues/487
-        if (manifest?.metadata?.creationTimestamp) {
-          manifest.metadata.creationTimestamp = new Date(manifest.metadata.creationTimestamp);
+
+      const validSpecs = manifests.filter(s => s?.kind);
```

Comment:
in the referenced link they were also filtering with ` && s.metadata` but it has been dropped

Reply:
add cast `as {kind: unknown}[]` at the end so you don't need it

Reply:
I struggled with this. The reason for filtering out objects without kind or metadata is 'don't bother to send objects that aren't valid Kubernetes YAML and likely to fail in patch/create', but our old create methods did this, and many of the tests are written w/o manifests.

If I add it back then the API of createResources() changes more, as some % of calls could 'succeed' (to do nothing) where it would have failed before. Alternately, applyResources() returns the successfully created resources so you can at least compare the count, but changing createResources() to return an array and updating all callers to check is yet another big PR/change.

Thoughts? I would like to add `&& s.metadata` back but was thinking that should be future PR.

Casting as `{kind:unknown}[]` caused lots of issues further down, so I eventually went with `KubernetesObject[]` instead. client.read() still doesn't like this and couldn't find another solution so any moved to there.